pyscf/solvent/pcm.py

```diff
@@ -0,0 +1,627 @@
+
+        de_dK = de_dS0 - fac * (de_dD + de_dA + de_dS1)
+        de += de_dR - de_dK
+    else:
+        raise RuntimeError(f"Unknown implicit solvent model: {pcmobj.method}")
+    
+    return de
+        
+def make_grad_object(grad_method):
+    '''
+    return solvent gradient object
+    '''
+    grad_method_class = grad_method.__class__
+    class WithSolventGrad(grad_method_class):
+        def __init__(self, grad_method):
+            self.__dict__.update(grad_method.__dict__)
+            self.de_solvent = None
+            self.de_solute = None
+            self._keys = self._keys.union(['de_solvent', 'de_solute'])
+
+        def kernel(self, *args, dm=None, atmlst=None, **kwargs):
+            dm = kwargs.pop('dm', None)
+            if dm is None:
+                dm = self.base.make_rdm1(ao_repr=True)
+            
+            self.de_solvent = grad_kernel(self.base.with_solvent, dm)
+            self.de_solute = grad_method_class.kernel(self, *args, **kwargs)
+            self.de = self.de_solute + self.de_solvent
+            
+            if self.verbose >= logger.NOTE:
+                logger.note(self, '--------------- %s (+%s) gradients ---------------',
+                            self.base.__class__.__name__,
+                            self.base.with_solvent.__class__.__name__)
+                logger.note(self, '----------------------------------------------')
+            return self.de
+
+        def _finalize(self):
+            # disable _finalize. It is called in grad_method.kernel method
+            # where self.de was not yet initialized.
+            pass
+
+    return WithSolventGrad(grad_method)
+
+class PCM(ddcosmo.DDCOSMO):
+    def __init__(self, mol):
+        ddcosmo.DDCOSMO.__init__(self, mol)
+        self.method = 'C-PCM'
+        self.vdw_scale = 1.2 # default value in qchem
+        self.surface = {}
+        self._intermediates = {}
+
+    def dump_flags(self, verbose=None):
+        logger.info(self, '******** %s (In testing) ********', self.__class__)
+        logger.warn(self, 'ddPCM is an experimental feature. It is '
+                    'still in testing.\nFeatures and APIs may be changed '
+                    'in the future.')
+        logger.info(self, 'lebedev_order = %s (%d grids per sphere)',
+                    self.lebedev_order, gen_grid.LEBEDEV_ORDER[self.lebedev_order])
+        logger.info(self, 'lmax = %s'         , self.lmax)
+        logger.info(self, 'eta = %s'          , self.eta)
+        logger.info(self, 'eps = %s'          , self.eps)
+        logger.info(self, 'frozen = %s'       , self.frozen)
+        logger.info(self, 'equilibrium_solvation = %s', self.equilibrium_solvation)
+        logger.debug2(self, 'radii_table %s', self.radii_table)
+        if self.atom_radii:
+            logger.info(self, 'User specified atomic radii %s', str(self.atom_radii))
+        self.grids.dump_flags(verbose)
+        return self
+
```

Comment:
When molecule object is changed, the surface should be updated accordingly, right? Here a `reset` method should be add to clear up the contents of `surface`, and also recursively call the `super().reset()` .